File: cal_view_app/cal_view_app/page/calendar_view_schedu/calendar_view_schedu.js

{% include "cal_view_app/cal_view_app/page/calendar_view_schedu/jquery.stickytable.js" %}


frappe.pages["calendar-view-schedu"].on_page_load = function (wrapper) {
	frappe.important_links = new frappe.ImpLinks(wrapper);
}


frappe.ImpLinks= Class.extend({

	init: function (wrapper) {
		this.page = frappe.ui.make_app_page({
			parent: wrapper,
			title: "Calendar View Schedule",
			single_column: true
		});
		this.wrapper = wrapper
		this.filters = {}
		this.make()
		this.get_imp_link_data()
    this.add_filters()
    $('[data-fieldname="new_task"]').css({"background-color":"#00b8ff"});
    this.add_menu_buttons()



	},

	make: function() {
		var me = this;
    $(`<div class="frappe-list list-container" > </div>`).appendTo(me.page.main);
	},

	get_imp_link_data:function(){
    var me = this
    
    $('.frappe-list').html("")

    var filters = {"project":me.project,"status":me.status,"date":me.date,"caliber":me.caliber, "department":me.department}
    frappe.call({
        "method": "cal_view_app.cal_view_app.page.calendar_view_schedu.calendar_view_schedu.get_imp_link_data",
        args: {
            filters: filters,
          },
        callback: function (r) {
          var html = r.message.html

          $('.frappe-list').html(html)
          me.bind_events(me)
          me.edit_task(me)
          me.delete_task(me)
          

        }//calback end
      })
	},

  add_filters:function(){
      var me = this
      user = frappe.session.user;
     
    
        me.page.add_field({
          fieldtype: 'Link',
          label: __('Project'),
          fieldname: 'project',
          options: "Project",
          onchange: function() {
            me.project = this.value?this.value:null
            me.get_imp_link_data()
          }
        })
      me.page.add_field({
          fieldtype: 'Select',
          label: __('Status'),
          fieldname: 'status',
          options: ["","Open","Working","Pending Review","Overdue","Completed","Cancelled"],
          default:"",
          onchange: function() {
            me.status = this.value?this.value:null
            me.get_imp_link_data()
          }
        })

      me.page.add_field({
          fieldtype: 'Link',
          label: __('Caliber'),
          fieldname: 'caliber',
          options: "User",
          onchange: function() {
            me.caliber = this.value?this.value:null
            me.get_imp_link_data()
          }
        })
      const today = frappe.datetime.get_today();
      me.page.add_field({
          fieldtype: 'Date',
          label: __('Date'),
          fieldname: 'date',
          default:today,
          reqd:1,
          onchange: function() {
            me.date = this.value?this.value:null
            me.get_imp_link_data()

          }
        })
       
          me.page.add_field({
          fieldtype: 'Link',
          label: __('Department'),
          fieldname: 'department',
          options:"Department",
          default:me.get_dep_default(),
          onchange:function() {
            me.department = this.value?this.value:null
            me.get_imp_link_data()
          }
        })

      me.page.add_field({
          fieldtype: 'Select',
          label: __('Assign Task'),
          fieldname: 'assign_task',
          options:["Assign Task", "Edit Task","Delete Task"],
          default: "Assign Task",
          onchange:function() {
            me.assign_task = this.value?this.value:null
          }

        })

      // me.page.add_field({
      //     fieldtype: 'Button',
      //     label: __('New Task'),
      //     fieldname: 'new_task',
      //     click:function() {
      //       console.log("qqq")
      //       me.new_task(me)
      //     }

      //   })

  },
  get_dep_default:function(){

    var department = ''
    frappe.call({
        method: "frappe.client.get_value",
        args: {
          doctype: "Employee",
          fieldname: "department",
          filters: { user_id: frappe.session.user },
        },
        async: false,
        callback: function(r) {
          if(r.message) {
            department = r.message.department
            
          }
        }
      });

    return department
  },
  add_menu_buttons:function(){
    me = this
    me.page.show_menu()

    me.page.add_menu_item(__("Refresh"), function() {
        me.get_imp_link_data()
    }, true)
    
    // me.page.add_menu_item(__("New Task"), function() {
    //     var new_doc = frappe.model.get_new_doc("Task");
    //     frappe.ui.form.make_quick_entry("Task", null, null, new_doc);
    // }, true)

    me.page.add_menu_item(__("Task List"), function() {
        window.location.href = 'desk#List/Task/List'
    }, true)

  },
  bind_events:function(me){
    var me = this;
    $('.assign_task').click(function() {
      // var phase = $(this).attr('data-phase')
      var assign_task = frappe.pages['calendar-view-schedu'].page.fields_dict.assign_task.get_value()

      var exceptional_msg = "{{ no such element: str object['task_name'] }}"
      
      var task_name = $(this).attr("task_name")
      var caliber_user = $(this).attr("caliber_user")

      if ((assign_task=="Assign Task") && (task_name) && (task_name != exceptional_msg) ){  
                      
              const assign_to = new frappe.ui.form.AssignToDialog({
                obj: this,
                method: 'frappe.desk.form.assign_to.add',
                doctype: "Task",
                docname: task_name,

              });

              assign_to.dialog.clear();
              assign_to.dialog.show();
              assign_to.dialog.set_value("assign_to",[caliber_user])
      }
      
      
    })




    $( ".dbclick" ).dblclick(function() {

      $(this).attr("contenteditable", "true");

      $('[contenteditable=true]').focus(function() {
              $(this).data("initialText", $(this).html());
          }).blur(function() {
              
              if ($(this).data("initialText") !== $(this).html()) {
                
                  var caliber = $(this).attr("caliber_user");
                  var project = $(this).closest('tr').find('td:first').text()
                  var col_index = $(this).index();
                  var col_index = col_index -1
                  var subject = $(this).html()
                  var date = $($("#main_tbl > thead > tr")[2]).find('th:eq('+col_index+')').text()

                  frappe.call({
                    method: "cal_view_app.cal_view_app.page.calendar_view_schedu.calendar_view_schedu.create_new_task",
                    args: {data:{
                      caliber: caliber,
                      project: project,
                      date: date,
                      filters: me.date,
                      subject: subject},
                    },
                    callback: function(r) {
                      if(r.message) {
                        // me.get_imp_link_data()
                      } 
                    }
                  });            

              }
            $(this).attr("contenteditable", "false");
          });
    });

  },


  edit_task: function(me){
    $('.assign_task').click(function() {
      var task_name = $(this).attr("task_name")

      var project = $(this).closest('tr').find('td:first').text()
      var exceptional_msg = "{{ no such element: str object['task_name'] }}"
      var assign_task = frappe.pages['calendar-view-schedu'].page.fields_dict.assign_task.get_value()
      if ((assign_task == "Edit Task") && (task_name) && ( task_name !=exceptional_msg)){
        var d = new frappe.ui.Dialog({
          title: __(task_name),
          fields: [
            {
              "label": "Subject",
              "fieldname": "subject",
              "fieldtype": "Data",

            },
            {
              "label": "Project",
              "fieldname": "project",
              "fieldtype": "Link",
              "options":"Project",
              "default":project


            },
            {
              "label": "Expected Start Date",
              "fieldname": "exp_start_date",
              "fieldtype": "Date",
            },
            {
              "label": "Department",
              "fieldname": "department",
              "fieldtype": "Link",
              "options":"Department"
            },
            
            {"fieldtype":"Column Break"},
            {
              "label": "Status",
              "fieldname": "status",
              "fieldtype": "Select",
              "default":"Open",
              "options":["Open",'Working','Pending Review','Overdue','Completed', 'Cancelled']
            },
            {
              "label": "Expected End Date",
              "fieldname": "exp_end_date",
              "fieldtype": "Date"
            },
            {
              "label": "Color",
              "fieldname": "color",
              "fieldtype": "Color"
            },
            {
              "label": "Completed By",
              "fieldname": "completed_by",
              "fieldtype": "Link",
              "options":"User"
            },
            
            {"fieldtype":"Section Break"},
            {
              "label": "Description",
              "fieldname": "description",
              "fieldtype": "Text Editor"
            },

          ],
          primary_action: function() {
            var data = d.get_values();
            frappe.call({
              method: "cal_view_app.cal_view_app.page.calendar_view_schedu.calendar_view_schedu.update_task_details",
              args: {
                data: data,
                task: task_name
              },
              callback: function(r) {
                if(r.message) {
                } 
              }
            });
            d.hide();
          },
          primary_action_label: __('Update')
        });
        me.get_dialog_default(d, task_name)
        d.show();
      }
    })

  },//edit task close
  get_dialog_default:function(d,task_name){
    
    var val = ''
    frappe.call({
        method: "cal_view_app.cal_view_app.page.calendar_view_schedu.calendar_view_schedu.get_task_details",
        args: {name: task_name},
        async: false,
        callback: function(r) {
          if(r.message) {
            // val =r.message[fld]
            d.set_values(r.message[0])
            
          }
        }
      });

    return
  },

  new_task: function(me){
    var new_doc = frappe.model.get_new_doc("Task");
    frappe.ui.form.make_quick_entry("Task", null, null, new_doc);
  },

  delete_task: function(me){
    var me = this;
    $('.assign_task').click(function() {
      var task_name = $(this).attr("task_name")
      var exceptional_msg = "{{ no such element: str object['task_name'] }}"
      var assign_task = frappe.pages['calendar-view-schedu'].page.fields_dict.assign_task.get_value()
      if ((assign_task == "Delete Task") && (task_name) && ( task_name !=exceptional_msg)){
        frappe.confirm(__("Do you really want to delete this task?"), function () {
          frappe.call({
            method: "cal_view_app.cal_view_app.page.calendar_view_schedu.calendar_view_schedu.delete_task",
            args: {
              "task_name": task_name
            },
            callback: function(r) {
            }
          })
        })
      }
    });
  }
})//class close


function change_next_week(me){

  var filter_date =  frappe.pages['calendar-view-schedu'].page.fields_dict.date.get_value()
  var curent_week_date = new Date(filter_date)
  var next_week_date = curent_week_date.setDate(curent_week_date.getDate() + 7)
  var next_week_date = new Date(next_week_date)
  console.log(next_week_date)
  frappe.pages['calendar-view-schedu'].page.fields_dict.date.set_value(next_week_date)

}

function change_previous_week(me){

  var filter_date =  frappe.pages['calendar-view-schedu'].page.fields_dict.date.get_value()
  var curent_week_date = new Date(filter_date)
  var next_week_date = curent_week_date.setDate(curent_week_date.getDate() - 7)
  var next_week_date = new Date(next_week_date)
  console.log(next_week_date)
  frappe.pages['calendar-view-schedu'].page.fields_dict.date.set_value(next_week_date)

}

File: cal_view_app/cal_view_app/page/calendar_view_schedu/calendar_view_schedu.py
from __future__ import unicode_literals
import frappe
from frappe.model.document import Document

from datetime import timedelta,date
import datetime
import calendar
import json
import time
import copy



def get_week_dates(base_date, start_day, end_day=None):
    """
    Return entire week of dates based on given date limited by start_day and end_day.
    If end_day is None, return only start_day.

    >>> from datetime import date
    >>> get_week_dates(date(2015,1,16), 3, 5)
    [datetime.date(2015, 1, 14), datetime.date(2015, 1, 15), datetime.date(2015, 1, 16)]

    >>> get_week_dates(date(2015,1,15), 2, 5)
    [datetime.date(2015, 1, 13), datetime.date(2015, 1, 14), datetime.date(2015, 1, 15), datetime.date(2015, 1, 16)]
    """

    
    if base_date.isoweekday() == 7 :
    	""" consider sunday as staring of week """

    	base_date += timedelta(days=1)

    monday = base_date - timedelta(days=base_date.isoweekday())
    week_dates = [monday + timedelta(days=i) for i in range(7)]

    return week_dates[start_day - 1:end_day or start_day]



@frappe.whitelist()
def get_imp_link_data(filters=None):
	filters	= json.loads(filters)

	week_days = ['Sunday', 'Monday', 'Tuesday','Wednesday', 'Thursday', 'Friday', 'Saturday']
	week_dates = get_week_dates(date.today(), 1, 7)
	if filters.get("date"):
		filter_date = datetime.datetime.strptime(filters.get("date"), '%Y-%m-%d').date()
		week_dates = get_week_dates(filter_date, 1, 7)


	week_start_date = week_dates[0]
	week_end_date = week_dates[-1]

	data = frappe.db.sql(""" select subject,project,exp_start_date,
				exp_end_date,caliber,completed_by, color ,name
						from `tabTask` 
				where {0} and exp_start_date IS NOT NULL and exp_end_date IS NOT NULL
				and (exp_start_date >= '{1}' and exp_end_date <= '{2}')
				ORDER BY caliber ASC""".format(get_filters_codition(filters),week_start_date,week_end_date),
				as_dict=1)
	


	template_data=[]

	for d in data:
		task_spils = []
		for dt in week_dates:
			if d.get("exp_start_date") <= dt <= d.get("exp_end_date"):
				task_spils.append(calendar.day_name[dt.weekday()])


		week_day = calendar.day_name[d.get("exp_start_date").weekday()]
		
		task_cell = {"subject":d.get("subject"),
					"color":d.get('color'),
					"task_name":d.get("name")}

		
		_p_line = {d:"" for d in week_days}
		
		_p_line.update({"project":d.get("project"),
				"caliber":d.get("caliber"), 
				"caliber_user":d.get("completed_by")
				})


		for t in task_spils:
			_p_line.update({t:task_cell})

		template_data.append(_p_line)



	_template_data = []
	week_days = ['Sunday', 'Monday', 'Tuesday','Wednesday', 'Thursday', 'Friday', 'Saturday']
	for td in template_data:
		pc = {"project":td.get("project"),"caliber_user":td.get("caliber_user")}
		
		if not _template_data:
			_template_data.append(td)
		
		else:
			project_line_found =0
			
			for _td in _template_data:
				
				_pc = {"project":_td.get("project"),"caliber_user":_td.get("caliber_user")}
				
				blank_day = [True for w in week_days if _td.get(w)]

				if pc == _pc and any(blank_day):

					for w in week_days:
						if not _td.get(w):
							
							_td.update({w:td.get(w)})
							td.update({w:""})

							project_line_found = 1
						else:
							black_p_line_to_add = 1


			if not project_line_found or black_p_line_to_add:
				blank_day = [True for w in week_days if td.get(w)]
				if any(blank_day):
					_template_data.append(td)


	template_data = _template_data

	if not template_data and filters.get('project') and filters.get('caliber'):
		caliber = frappe.db.get_value("User", {'name':filters.get('caliber')}, 'full_name')
		template_data.append({'Sunday': '', 'Monday': '', 'Tuesday': '', 'Wednesday': '', 'Thursday': '', 'Friday': '', 'Saturday': '', 'project': filters.get('project'), 'caliber': caliber, 'caliber_user': filters.get('caliber')})


	week_dates = [w.strftime("%d/%b") for w in week_dates]
	
	path = 'cal_view_app/cal_view_app/page/calendar_view_schedu/calendar_view_schedu.html'

	html=frappe.render_template(path,{'data':{"week_days":week_dates,
											"template_data":template_data, "department":filters.get('department')}})

	return {"html":html}

def get_filters_codition(filters):
	conditions = "1=1"
	if filters.get("project"):
		conditions += " and project = '{0}'".format(filters.get('project'))
	if filters.get("status"):
		conditions += " and status = '{0}'".format(filters.get('status'))
	if filters.get("caliber"):
		conditions += " and completed_by = '{0}'".format(filters.get('caliber'))
	if filters.get("department"):
		conditions += " and department = '{0}'".format(filters.get('department'))


	return conditions

@frappe.whitelist()
def update_task_details(data, task):
	task_data=json.loads(data)

	doc = frappe.get_doc("Task", task)
	doc.subject = task_data.get('subject')
	doc.status = task_data.get('status')
	doc.exp_start_date = task_data.get('exp_start_date')
	doc.exp_end_date = task_data.get('exp_end_date')
	doc.department = task_data.get('department')
	doc.description = task_data.get('description')
	doc.color = task_data.get('color')
	doc.project = task_data.get('project')
	doc.save()

@frappe.whitelist()
def get_task_details(name):

	task_details = frappe.db.sql(""" select subject, status, exp_start_date, exp_end_date, department, description, completed_by, color	from `tabTask` where name = '{0}'""".format(name),as_dict=1)

	return task_details


@frappe.whitelist()
def create_new_task(data):
	task_data = json.loads(data)
	split_date=task_data.get('date').split("/")
	task_date = split_date[1]+"-"+split_date[0]+"-"+task_data.get('filters')[0:4]
	exp_start_date = datetime.datetime.strptime(task_date, '%b-%d-%Y')

	doc=frappe.new_doc("Task")
	doc.subject = task_data.get('subject')
	doc.project = task_data.get('project')
	doc.completed_by = task_data.get('caliber')
	doc.exp_start_date = exp_start_date
	doc.exp_end_date = exp_start_date
	doc.save()
	frappe.db.commit()

	return True



@frappe.whitelist()
def delete_task(task_name):
	frappe.db.sql("""delete from `tabTask`
			where name='{0}'""".format(task_name))
	frappe.db.commit()

	return True